components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java

```diff
@@ -81,9 +84,18 @@ public void onMessage(Message message) {
                              * timestamp - system time of the event published
                              * event - event data
                              */
-                            handleNotificationMessage((String) map.get(APIConstants.EVENT_TYPE),
-                                    (Long) map.get(APIConstants.EVENT_TIMESTAMP),
-                                    (String) map.get(APIConstants.EVENT_PAYLOAD));
+
+                            ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+                            final Runnable task = new Runnable() {
+                                @Override
+                                public void run() {
+                                    handleNotificationMessage((String) map.get(APIConstants.EVENT_TYPE),
```

Comment:
I think putting it this way would make each and every event handled in a different thread. And be started with a delay. Can we only start a task when an API is need to be deployed.

Reply:
We can do the change like below:
1. Create a new ThreadScheduler at the JMSMessageListener with a fixed number of threads.
2. While submitting the Task to the above Scheduler override the run method either with deploy or undeploy methods.
3. Use no delay while submitting the task - we initially discussed about it, but in a later discussion changed the approach.
4. Now the thread will be different from JMS event receiving thread. So do the waiting in the same thread by using a thread sleep and continue with the pulling and deployment.

Reply:
https://github.com/wso2/carbon-apimgt/pull/9016/commits/67da6203b4479b5838882a8416f61db2eb3af7fe